tests/catch/include/threaded_zig_zag_test.hh

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright (c) 2022 Advanced Micro Devices, Inc. All rights reserved.
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#pragma once
+
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+/*
+Guarantees total ordering between parent and child thread
+PARENT      CHILD
+THREAD      THREAD
+TestPart1
+         \
+          \
+           \
+            TestPart2
+           /
+          /
+         /
+TestPart3
+         \
+          \
+           \
+            TestPart4
+Usage:
+Define a derived class which inherits from ThreadedZigZagTest instantiated with that selfsame class,
+which implements the appropriate test methods
+class DerivedTestClass : public ThreadedZigZagTest<DerivedTestClass> {
+  void TestPart1() {...}
+  void TestPart2() {...}
+  void TestPart3() {...}
+  void TestPart4() {...}
+};
+The derived class can contain state that the test requires.
+*/
+
+template <typename T> class ThreadedZigZagTest {
+ public:
+  void run() {
+    // 1.
+    static_cast<T*>(this)->TestPart1();
+
+    auto t = std::thread([this] {
+      // 2.
+      static_cast<T*>(this)->TestPart2();
+
+      {
+        std::lock_guard<std::mutex> lock(mtx_);
+        ready_ = true;
+      }
+      cv_.notify_one();
+
+      {
+        std::unique_lock<std::mutex> lock(mtx_);
+        cv_.wait(lock, [this] { return !ready_; });
+      }
+
+      // 4.
+      static_cast<T*>(this)->TestPart4();
+    });
+
+    {
+      std::unique_lock<std::mutex> lock(mtx_);
+      cv_.wait(lock, [this] { return ready_; });
+    }
+
+    // 3.
+    static_cast<T*>(this)->TestPart3();
+
+    {
+      std::lock_guard<std::mutex> lock(mtx_);
+      ready_ = false;
+    }
+    cv_.notify_one();
+
+    // Finalize
+    t.join();
+    HIP_CHECK_THREAD_FINALIZE();
+  }
+
+  void TestPart1() const {}
+  void TestPart2() const {}
+  void TestPart3() const {}
+  void TestPart4() const {}
```

Comment:
Don't they have to be virtual?

Reply:
No. These are only default implementations in case a derived class doesn't need to implement one of the methods. If a derived class implements a method of the same name as one of the above it will shadow the parent class implementation and will be called instead. Virtual dispatch was not intended and is not necessary.

Reply:
I get it. Because it's static_cast<T*> that it doesn't need to be virtual.